include/gsl/span

```diff
@@ -104,156 +105,187 @@ namespace details
-    struct is_std_array : public is_std_array_oracle<std::remove_cv_t<T>>
+    struct is_std_array : is_std_array_oracle<std::remove_cv_t<T>>
     {
     };
 
-    template <std::ptrdiff_t From, std::ptrdiff_t To>
+    template <std::size_t From, std::size_t To>
     struct is_allowed_extent_conversion
-        : public std::integral_constant<bool, From == To || From == gsl::dynamic_extent ||
-                                                  To == gsl::dynamic_extent>
+        : std::integral_constant<bool, From == To || To == gsl::dynamic_extent>
     {
     };
 
     template <class From, class To>
     struct is_allowed_element_type_conversion
-        : public std::integral_constant<bool, std::is_convertible<From (*)[], To (*)[]>::value>
+        : std::integral_constant<bool, std::is_convertible<From (*)[], To (*)[]>::value>
     {
     };
 
-    template <class Span, bool IsConst>
+    template <class Type>
     class span_iterator
     {
-        using element_type_ = typename Span::element_type;
-
     public:
-#ifdef _MSC_VER
-        // Tell Microsoft standard library that span_iterators are checked.
-        using _Unchecked_type = typename Span::pointer;
-#endif
-
         using iterator_category = std::random_access_iterator_tag;
-        using value_type = std::remove_cv_t<element_type_>;
-        using difference_type = typename Span::index_type;
+        using value_type = std::remove_cv_t<Type>;
+        using difference_type = std::ptrdiff_t;
+        using pointer = Type*;
+        using reference = Type&;
 
-        using reference = std::conditional_t<IsConst, const element_type_, element_type_>&;
-        using pointer = std::add_pointer_t<reference>;
-
-        span_iterator() = default;
-
-        constexpr span_iterator(const Span* span, difference_type idx) noexcept
-            : span_(span), index_(idx)
-        {}
+#ifdef _MSC_VER
+        using _Unchecked_type = pointer;
+#endif // _MSC_VER
+        constexpr span_iterator() = default;
 
-        friend span_iterator<Span, true>;
-        template <bool B, std::enable_if_t<!B && IsConst>* = nullptr>
-        constexpr span_iterator(const span_iterator<Span, B>& other) noexcept
-            : span_iterator(other.span_, other.index_)
+        constexpr span_iterator(pointer begin, pointer end, pointer current)
+            : begin_(begin), end_(end), current_(current)
         {}
 
-        GSL_SUPPRESS(bounds.1) // NO-FORMAT: attribute
-        constexpr reference operator*() const
+        constexpr operator span_iterator<const Type>() const noexcept
         {
-            Expects(index_ != span_->size());
-            return *(span_->data() + index_);
+            return {begin_, end_, current_};
         }
 
-        constexpr pointer operator->() const
+        constexpr reference operator*() const noexcept
         {
-            Expects(index_ != span_->size());
-            return span_->data() + index_;
+            Expects(begin_ && current_ && end_);
```

Comment:
The following test lets you delete the current_ test. (If begin and end are not null and current is in between, then current can't be null either)

Occurs below.

Reply:
makes sense, updating.